nvflare/client/flare_agent.py

```diff
@@ -0,0 +1,356 @@
+        self.logger.info(f"{channel} pipe status changed to {msg.topic}: {msg.data}")
+        self.asked_to_stop = True
+        pipe_handler.stop()
+
+    def stop(self):
+        """Stop the agent. After this is called, there will be no more communications between CJ and agent.
+
+        Returns: None
+
+        """
+        self.asked_to_stop = True
+        self.pipe_handler.stop()
+
+    def shareable_to_task_data(self, shareable: Shareable):
+        """Convert the Shareable object received from the TaskExchanger to an app-friendly format.
+        Subclass can override this method to convert to its own app-friendly task data.
+        By default, we convert to DXO object.
+
+        Args:
+            shareable: the Shareable object received from the TaskExchanger.
+        """
+        try:
+            dxo = from_shareable(shareable)
+
+            # add training-related headers carried in the Shareable header to the DXO meta.
+            total_rounds = shareable.get_header(AppConstants.NUM_ROUNDS)
+            if total_rounds is not None:
+                dxo.set_meta_prop(MetaKey.TOTAL_ROUNDS, total_rounds)
+            current_round = shareable.get_header(AppConstants.CURRENT_ROUND)
+            if current_round is not None:
+                dxo.set_meta_prop(MetaKey.CURRENT_ROUND, current_round)
+            return dxo
+        except Exception as ex:
+            self.logger.error(f"failed to extract DXO from shareable object: {ex}")
+            raise ex
+
+    def get_task(self, timeout: Optional[float] = None):
+        """Get a task from FLARE. This is a blocking call.
+
+        Args:
+            timeout (float, optional): If specified, this call is blocked only for the specified amount of time.
+                If not specified, this call is blocked forever until a task has been received or agent has been closed.
+
+        Returns:
+            None if no task is available during before timeout; or a Task object if task is available.
+        Raises:
+            AgentClosed exception if the agent has been closed before timeout.
+            CallStateError exception if the call has not been made properly.
+            AgentAbortException: If the other endpoint of the pipe requests to abort.
+            AgentEndException: If the other endpoint has ended.
+            AgentPeerGoneException: If the other endpoint is gone.
+
+        Note: the application must make the call only when it is just started or after a previous task's result
+        has been submitted.
+
+        """
+        start_time = time.time()
+        while True:
+            if self.asked_to_stop:
+                raise AgentClosed("agent closed")
+
+            if self.current_task:
+                raise CallStateError("application called get_task while the current task is not processed")
+
+            if timeout is not None and time.time() - start_time >= timeout:
+                self.pipe_handler.notify_abort()
```

Comment:
Why should we notify_abort here? This is the case that the caller only wants to wait for some time, it doesn't mean that the caller is aborting.

Reply:
Make sense, let me fix it.